internal/app/go-filecoin/internal/submodule/discovery_submodule.go

```diff
@@ -0,0 +1,111 @@
+package submodule
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-filecoin/internal/pkg/block"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/config"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/discovery"
+	"github.com/filecoin-project/go-filecoin/internal/pkg/net"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	"github.com/pkg/errors"
+)
+
+var log = logging.Logger("node") // nolint: deadcode
+
+// DiscoverySubmodule enhances the `Node` with peer discovery capabilities.
+type DiscoverySubmodule struct {
+	Bootstrapper *discovery.Bootstrapper
+
+	// PeerTracker maintains a list of peers.
+	PeerTracker *discovery.PeerTracker
+
+	// HelloHandler handle peer connections for the "hello" protocol.
+	HelloHandler *discovery.HelloProtocolHandler
+}
+
+type discoveryConfig interface {
+	GenesisCid() cid.Cid
+}
+
+// NewDiscoverySubmodule creates a new discovery submodule.
+func NewDiscoverySubmodule(ctx context.Context, config discoveryConfig, bsConfig *config.BootstrapConfig, network *NetworkSubmodule) (DiscoverySubmodule, error) {
+	periodStr := bsConfig.Period
+	period, err := time.ParseDuration(periodStr)
+	if err != nil {
+		return DiscoverySubmodule{}, errors.Wrapf(err, "couldn't parse bootstrap period %s", periodStr)
+	}
+
+	// bootstrapper maintains connections to some subset of addresses
+	ba := bsConfig.Addresses
+	bpi, err := net.PeerAddrsToAddrInfo(ba)
+	if err != nil {
+		return DiscoverySubmodule{}, errors.Wrapf(err, "couldn't parse bootstrap addresses [%s]", ba)
+	}
+
+	minPeerThreshold := bsConfig.MinPeerThreshold
+
+	// create a bootstrapper
+	bootstrapper := discovery.NewBootstrapper(bpi, network.PeerHost, network.PeerHost.Network(), network.Router, minPeerThreshold, period)
+
+	// set up peer tracking
+	peerTracker := discovery.NewPeerTracker(network.PeerHost.ID())
+
+	return DiscoverySubmodule{
+		Bootstrapper: bootstrapper,
+		PeerTracker:  peerTracker,
+		HelloHandler: discovery.NewHelloProtocolHandler(network.PeerHost, config.GenesisCid(), network.NetworkName),
+	}, nil
+}
+
+type discoveryNode interface {
+	Network() NetworkSubmodule
+	Chain() ChainSubmodule
+}
+
+// Start starts the discovery submodule for a node.
+func (m *DiscoverySubmodule) Start(node discoveryNode) error {
+	// Start bootstrapper.
+	m.Bootstrapper.Start(context.Background())
+
+	// Register peer tracker disconnect function with network.
+	m.PeerTracker.RegisterDisconnect(node.Network().PeerHost.Network())
+
+	// Start up 'hello' handshake service
+	peerDiscoveredCallback := func(ci *block.ChainInfo) {
+		m.PeerTracker.Track(ci)
+		err := node.Chain().SyncDispatch.SendHello(ci)
+		if err != nil {
+			log.Errorf("error receiving chain info from hello %s: %s", ci, err)
+			return
+		}
+		// For now, consider the initial bootstrap done after the syncer has (synchronously)
+		// processed the chain up to the head reported by the first peer to respond to hello.
```

Comment:
@ZenGround0 does `SyncDispatch.SendHello` fetch and validate synchronously? If not, then this comment and the rationale of the ChainSynced latch is invalid.